File: DP(Dynamic_Programming, 동적계획법)/백준_2747_피보나치 수/피보나치수.java
import java.io.*;
import java.util.*;

public class Main {
    static int[] d = new int[50];

    public static int dp(int n){
        //1. base check
        if(n <= 1) return n;

        //2. memoization
        if(d[n] > 0) return d[n];

        //3. recurrence relation
        d[n] = dp(n-1) + dp(n-2);

        return d[n];
    }

    public static void main(String[] args) throws IOException {
        Scanner s = new Scanner(System.in);
        int n = s.nextInt();
        System.out.println(dp(n));

    }
}

File: DP(Dynamic_Programming, 동적계획법)/백준_2747_피보나치 수/피보나치수.kt
import java.util.*

val d = IntArray(50, {0})

fun dp(n: Int):Int {
    if(n <= 1) return n

    if(d[n] > 0) return d[n]

    d[n] = dp(n-1) + dp(n-2)

    return d[n]
}

fun main(args: Array<String>) {
    val s = Scanner(System.`in`)
    val input = s.nextInt()
    print("${dp(input)}")
}